# **Pandas**
Pandas is a Python library.

Pandas is used to analyze data. What is Pandas? Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data. Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

In [ ]:
#Import Pandas
import pandas
#Create an alias with the as keyword while importing.
import pandas as pd
#Now the Pandas package can be referred to as pd instead of pandas.

# **Pandas Series**
A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [ ]:
#Create a simple Pandas Series from a list:
import pandas as pd
a = [1, 7, 2]
myvar = pd.Series(a)
print(myvar)

0    1
1    7
2    2
dtype: int64


In [ ]:
# Labels
# If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

# This label can be used to access a specified value.

# Example
# Return the first value of the Series:

print(myvar[0])

1


In [ ]:
# Key/Value Objects as Series
# You can also use a key/value object, like a dictionary, when creating a Series.

# Example
# Create a simple Pandas Series from a dictionary:

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)

day1    420
day2    380
day3    390
dtype: int64


In [ ]:
# To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.

# Example
# Create a Series using only data from "day1" and "day2":

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


# **DataFrames**
Data sets in Pandas are usually multi-dimensional tables, called DataFrames. Series is like a column, a DataFrame is the whole table.

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [ ]:
# Example
# Create a DataFrame from two Series:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45


# **Pandas Read CSV**

In [ ]:
# Load Files Into a DataFrame
# If your data sets are stored in a file, Pandas can load them into a DataFrame.

# Example
# Load a comma separated file (CSV file) into a DataFrame:

import pandas as pd

df = pd.read_csv('Data.csv')

print(df)

     symboling  normalized-losses         make  ... horsepower-binned diesel gas
0            3                122  alfa-romero  ...            Medium      0   1
1            3                122  alfa-romero  ...            Medium      0   1
2            1                122  alfa-romero  ...            Medium      0   1
3            2                164         audi  ...            Medium      0   1
4            2                164         audi  ...            Medium      0   1
..         ...                ...          ...  ...               ...    ...  ..
196         -1                 95        volvo  ...            Medium      0   1
197         -1                 95        volvo  ...              High      0   1
198         -1                 95        volvo  ...            Medium      0   1
199         -1                 95        volvo  ...            Medium      1   0
200         -1                 95        volvo  ...            Medium      0   1

[201 rows x 29 columns]


In [ ]:
#Read CSV files
# Example
# Load the CSV into a DataFrame:

import pandas as pd

df = pd.read_csv('Data.csv')

print(df.to_string())

     symboling  normalized-losses           make aspiration num-of-doors   body-style drive-wheels engine-location  wheel-base    length     width  height  curb-weight engine-type num-of-cylinders  engine-size fuel-system      bore  stroke  compression-ratio  horsepower     peak-rpm  city-mpg  highway-mpg    price  city-L/100km horsepower-binned  diesel  gas
0            3                122    alfa-romero        std          two  convertible          rwd           front        88.6  0.811148  0.890278    48.8         2548        dohc             four          130        mpfi  3.470000    2.68               9.00  111.000000  5000.000000        21           27  13495.0     11.190476            Medium       0    1
1            3                122    alfa-romero        std          two  convertible          rwd           front        88.6  0.811148  0.890278    48.8         2548        dohc             four          130        mpfi  3.470000    2.68               9.00  111.000000  5000.0

In [ ]:
# If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:

# Example
# Print the DataFrame without the to_string() method:

import pandas as pd

df = pd.read_csv('Data.csv')

print(df)

     symboling  normalized-losses         make  ... horsepower-binned diesel gas
0            3                122  alfa-romero  ...            Medium      0   1
1            3                122  alfa-romero  ...            Medium      0   1
2            1                122  alfa-romero  ...            Medium      0   1
3            2                164         audi  ...            Medium      0   1
4            2                164         audi  ...            Medium      0   1
..         ...                ...          ...  ...               ...    ...  ..
196         -1                 95        volvo  ...            Medium      0   1
197         -1                 95        volvo  ...              High      0   1
198         -1                 95        volvo  ...            Medium      0   1
199         -1                 95        volvo  ...            Medium      1   0
200         -1                 95        volvo  ...            Medium      0   1

[201 rows x 29 columns]


In [ ]:
#Load the JSON file into a DataFrame:

import pandas as pd

df = pd.read_json('Data.json')

print(df.to_string())

ValueError: ignored

# **Viewing the Data**
One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

In [ ]:
#Get a quick overview by printing the first 10 rows of the DataFrame:

import pandas as pd

df = pd.read_csv('Data.csv')

print(df.head(10))

   symboling  normalized-losses         make  ... horsepower-binned diesel gas
0          3                122  alfa-romero  ...            Medium      0   1
1          3                122  alfa-romero  ...            Medium      0   1
2          1                122  alfa-romero  ...            Medium      0   1
3          2                164         audi  ...            Medium      0   1
4          2                164         audi  ...            Medium      0   1
5          2                122         audi  ...            Medium      0   1
6          1                158         audi  ...            Medium      0   1
7          1                122         audi  ...            Medium      0   1
8          1                158         audi  ...            Medium      0   1
9          2                192          bmw  ...               Low      0   1

[10 rows x 29 columns]


In [ ]:
# Note: if the number of rows is not specified, the head() method will return the top 5 rows.

# Example
# Print the first 5 rows of the DataFrame:

import pandas as pd

df = pd.read_csv('Data.csv')

print(df.head())

   symboling  normalized-losses         make  ... horsepower-binned diesel gas
0          3                122  alfa-romero  ...            Medium      0   1
1          3                122  alfa-romero  ...            Medium      0   1
2          1                122  alfa-romero  ...            Medium      0   1
3          2                164         audi  ...            Medium      0   1
4          2                164         audi  ...            Medium      0   1

[5 rows x 29 columns]


In [ ]:
# There is also a tail() method for viewing the last rows of the DataFrame.

# The tail() method returns the headers and a specified number of rows, starting from the bottom.

# Example
# Print the last 5 rows of the DataFrame:

print(df.tail())

     symboling  normalized-losses   make  ... horsepower-binned diesel gas
196         -1                 95  volvo  ...            Medium      0   1
197         -1                 95  volvo  ...              High      0   1
198         -1                 95  volvo  ...            Medium      0   1
199         -1                 95  volvo  ...            Medium      1   0
200         -1                 95  volvo  ...            Medium      0   1

[5 rows x 29 columns]


In [ ]:
# Info About the Data
# The DataFrames object has a method called info(), that gives you more information about the data set.

# Example
# Print information about the data:


print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    int64  
 2   make               201 non-null    object 
 3   aspiration         201 non-null    object 
 4   num-of-doors       201 non-null    object 
 5   body-style         201 non-null    object 
 6   drive-wheels       201 non-null    object 
 7   engine-location    201 non-null    object 
 8   wheel-base         201 non-null    float64
 9   length             201 non-null    float64
 10  width              201 non-null    float64
 11  height             201 non-null    float64
 12  curb-weight        201 non-null    int64  
 13  engine-type        201 non-null    object 
 14  num-of-cylinders   201 non-null    object 
 15  engine-size        201 non-null    int64  
 16  fuel-system        201 non

# **Null Values**
The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data

# **Cleaning Data**
Data Cleaning Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells Data in wrong format Wrong data Duplicates

Empty Cells Empty cells can potentially give you a wrong result when you analyze data.

Remove Rows One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

In [ ]:
# Example
# Return a new Data Frame with no empty cells:

import pandas as pd

df = pd.read_csv('Data.csv')

new_df = df.dropna()

print(new_df.to_string())

     symboling  normalized-losses           make aspiration num-of-doors   body-style drive-wheels engine-location  wheel-base    length     width  height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  bore  stroke  compression-ratio  horsepower     peak-rpm  city-mpg  highway-mpg    price  city-L/100km horsepower-binned  diesel  gas
0            3                122    alfa-romero        std          two  convertible          rwd           front        88.6  0.811148  0.890278    48.8         2548        dohc             four          130        mpfi  3.47    2.68               9.00  111.000000  5000.000000        21           27  13495.0     11.190476            Medium       0    1
1            3                122    alfa-romero        std          two  convertible          rwd           front        88.6  0.811148  0.890278    48.8         2548        dohc             four          130        mpfi  3.47    2.68               9.00  111.000000  5000.000000       

In [ ]:
# Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

# If you want to change the original DataFrame, use the inplace = True argument:

# Example
# Remove all rows with NULL values:

import pandas as pd

df = pd.read_csv('Data.csv')

df.dropna(inplace = True)

print(df.to_string())

     symboling  normalized-losses           make aspiration num-of-doors   body-style drive-wheels engine-location  wheel-base    length     width  height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  bore  stroke  compression-ratio  horsepower     peak-rpm  city-mpg  highway-mpg    price  city-L/100km horsepower-binned  diesel  gas
0            3                122    alfa-romero        std          two  convertible          rwd           front        88.6  0.811148  0.890278    48.8         2548        dohc             four          130        mpfi  3.47    2.68               9.00  111.000000  5000.000000        21           27  13495.0     11.190476            Medium       0    1
1            3                122    alfa-romero        std          two  convertible          rwd           front        88.6  0.811148  0.890278    48.8         2548        dohc             four          130        mpfi  3.47    2.68               9.00  111.000000  5000.000000       

In [ ]:
#Replace Empty Values
# Another way of dealing with empty cells is to insert a new value instead.

# This way you do not have to delete entire rows just because of some empty cells.

# The fillna() method allows us to replace empty cells with a value:

# Example
# Replace NULL values with the number 130:

import pandas as pd

df = pd.read_csv('Data.csv')

df.fillna(130, inplace = True)

In [ ]:
# Replace Only For Specified Columns
# The example above replaces all empty cells in the whole Data Frame.

# To only replace empty values for one column, specify the column name for the DataFrame:

# Example
# Replace NULL values in the "Calories" columns with the number 130:

import pandas as pd

df = pd.read_csv('Data.csv')

df["Calories"].fillna(130, inplace = True)

# **Replace Using Mean, Median, or Mode**

In [ ]:
# A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

# Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

# Example
# Calculate the MEAN, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mean()

df["Calories"].fillna(x, inplace = True)

In [ ]:
# Example
# Calculate the MEDIAN, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].median()

df["Calories"].fillna(x, inplace = True)

In [ ]:
# Example
# Calculate the MODE, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mode()[0]

df["Calories"].fillna(x, inplace = True)

# **Removing duplicates**

In [ ]:
# Example
# Returns True for every row that is a duplicate, othwerwise False:
print(df.duplicated())

In [ ]:
# Removing Duplicates
# To remove duplicates, use the drop_duplicates() method.

# Example
# Remove all duplicates:

df.drop_duplicates(inplace = True)

#Remember: The (inplace = True) will make sure that the method does NOT return a new DataFrame, but it will remove all duplicates from the original DataFrame.


# **Finding Relationships**
A great aspect of the Pandas module is the corr() method.

The corr() method calculates the relationship between each column in your data set.

In [ ]:
# Example
# Show the relationship between the columns:

df.corr()
#Note: The corr() method ignores "not numeric" columns

Result Explained The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

What is a good correlation? It depends on the use, but I think it is safe to say you have to have at least 0.6 (or -0.6) to call it a good correlation.

Perfect Correlation: We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

Good Correlation: "Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

Bad Correlation: "Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.